### ISSUE TITLE
Visual Studio code does not recognize jest ts

### ISSUE THREAD

**Issue by panitaxx** [microsoft/TypeScript-Node-Starter#196]
I have installed this package but VScode complains that Jest methods do not exist (describe, it, expect). Any  way to fix this?

**Comment by peterblazejewicz** [comment]
Make sure your `package.json` contains an entry like `"@types/jest": "^24.0.15",`, reinstall NPM and use 'Typescript: Restart TS Server' option in VSCode. (CMD+SHIFT+P)

**Comment by mkolodziejczyk-branch** [comment]
And if it still does not work?

**Comment by dschuessler** [comment]
Try restarting. This sometimes help.

I also experience VSCode not recognizing types every once in a while though they are definitely installed. Sometimes restarting helps. Sometimes it doesn't. I wish I knew how to reproduce it. This clearly has been an issue for some while.

**Comment by mkolodziejczyk-branch** [comment]
Actually it somehow magically started working after a while... Thanks :)

**Comment by peterblazejewicz** [comment]
@mkolodziejczyk-branch nice to hear!

**Comment by JelsB** [comment]
The only way I was able to fix this, was by adding the `tests/` folder to "include" in the `tsconfig.json` file:

```
"include": [
  "src/**/*.ts",
  "tests/*.ts"
] 
```

**Comment by anan44** [comment]
I am having a similar issue. VSC recognises jest types if the project is opened from project root where tsconfig.json is located. If the project is opened from its parent directory, only way to get VSC to recognise the jest types is to include tests in tsconfig.json. Unfortunately this causes the tests to be compiled as well, which is often undesirable.

It would be great if VSC could recognise jest types even if opened in monorepo style.

**Comment by tomgallagher** [comment]
If you don't want to add the tests folder to the `tsconfig.json` - why would you? - then you can get rid of the annoying problem messages in VS Code by adding this to the top of the folder

`import { describe, test, expect } from '@jest/globals';`

And any others you might need. Still hard to understand why @types/jest are not recognised.

**Comment by boillodmanuel** [comment]
@JelsB , I also included `tests/**/*.ts` but this change the build. Indeed, `dist` dir contains `src` and `test` folder.

**Comment by dezoito** [comment]
> I am having a similar issue. VSC recognises jest types if the project is opened from project root where tsconfig.json is located. If the project is opened from its parent directory, only way to get VSC to recognise the jest types is to include tests in tsconfig.json. Unfortunately this causes the tests to be compiled as well, which is often undesirable.
> 

Thank you! This was driving me nuts until I read your post.

I had the following section in my tsconfig.json file, and commenting out "node_modules" fixed the issue... 

```
  "exclude": [
    // "node_modules",
    "**/mocks/*.ts",
    "**/*.spec.ts",
    "**/*.test.ts",
    "**/*.spec.tsx",
    "**/*.test.tsx"
  ],
```

**Comment by codingLogan** [comment]
This may be old, but I figure adding my details may be useful to anyone coming around looking for an answer.

_TLDR; Open the project folder directly, instead of opening a mono-repo's root directory_

I noticed that VSCode does not find the jest globals when a parent folder of a project is your current "root".  Example below is the structure I had, and while editing test1.test.ts the gobals _test_ and _expect_ are not found by VSCode even though @types/jest is installed.

parent-folder
- package
    - src
        - __tests__
            -  test1.test.ts
            - test2.test.ts
    - jestconfig.json
    - package.json
    - tsconfig.json
- test-client
    - etc...

If I open a new VSCode instance and choose to open the _package_ folder as the root, VSCode does find the jest globals and all looks and works great.

Hope this helps someone 👍

**Comment by jonahx** [comment]
@codingLogan Thanks, that solves the problem for me, but now how do you work on files in another branch off the root?

**Comment by codingLogan** [comment]
> @codingLogan Thanks, that solves the problem for me, but now how do you work on files in another branch off the root?

Hey @jonahx , in my particular case I opened  up a few instances of VS Code rather than having just one of the whole monorepo.  One was opened up with _package_ as the root, and one with _test-client_.
